File: python/basics/2-pointer/11_container_with_most_water.py
#https://leetcode.com/problems/container-with-most-water/description/

#2-pointer - move the pointer with lesser height
#time O(n)
#space O(1)
def container_with_most_water(height):
    result = 0
    #l h
    #min  & width

    N = len(height)
    l = 0
    r = N-1
    max_area = 0
    while r > l:
        min_height = min(height[l], height[r])
        area = min_height * (r-l)
        max_area = max(area, max_area)
        if height[l] >= height[r]:
            r -= 1
        elif height[l] < height[r]:
            l += 1
    return max_area




height = [1,8,6,2,5,4,8,3,7]
result = container_with_most_water(height)
print("result: " + str(result))
#Output: 49

#Explanation: The above vertical lines are represented by array [1,8,6,2,5,4,8,3,7]. In this case, the max area of water (blue section) the container can contain is 49.


height = [1,1]
result = container_with_most_water(height)
print("result: " + str(result))

#output 1

File: problem_solving/2-pointer/cpp/11_container_with_most_water.cpp
/*
#https://leetcode.com/problems/container-with-most-water/description/

#2-pointer - move the pointer with lesser height
#time O(n)
#space O(1)
*/

#include "../../../cpp/common_header.h"
#include "../../../cpp/utils/utils.h"


int container_with_most_water(const std::vector<int> heights) {
    int water = 0;
    int l = 0;
    int r = heights.size()-1;
    int max_area = 0;
    while(l < r) {
        int area = ( std::min(heights[l], heights[r]) * (r-l) );
        if(area > max_area) {
            max_area = area;
        }
        if(heights[l] <= heights[r]) {
            l++;
        } else {
            r--;
        }
    }
    return max_area;
}


int main() {
    std::vector<int> height = {1,8,6,2,5,4,8,3,7};
    int result = container_with_most_water(height);
    std::cout << "result: " << result << std::endl;
    //#Output: 49

    //#Explanation: The above vertical lines are represented by array [1,8,6,2,5,4,8,3,7]. In this case, the max area of water (blue section) the container can contain is 49.


    height = {1,1};
    result = container_with_most_water(height);
    std::cout << "result: " << result << std::endl;
    //#output 1

}
